### ISSUE TITLE
Parser doesn't like dates without separator with 4-digits years

### ISSUE THREAD

**Issue by xmo-odoo** [dateutil/dateutil#796]
I was looking at the various "fuzzy" parsers for dates, and dateutil.parser raises when parsing ddmmyyyy or mmddyyyy.

It also does not deal well with times above 48h, or seconds-only timestamps, but no parser handles these correctly.

Here's the full testcase table, the first two columns are the "kind" of pattern and the actual test date(time), the others are various parsers in their default configuration except for dateutil-fuzzy which is `partial(dateutil.parser.parse, fuzzy=True)`:

| pattern                      | original                   | babel                         | dateutil                   | dateutil-fuzzy             | dateparser                 |
|:-----------------------------|:---------------------------|:------------------------------|:---------------------------|:---------------------------|:---------------------------|
| dd/mm/yy                     | 03/08/06                   | 2006-08-03                    | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        |
| dd/mm/yyyy                   | 03/08/2006                 | 2006-08-03                    | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        |
| d/m/yy                       | 3/8/06                     | 2006-08-03                    | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        |
| d/m/yyyy                     | 3/8/2006                   | 2006-08-03                    | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        |
| ddmmyy                       | 030806                     | list index out of range       | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        | 2006-03-08 00:00:00        |
| ddmmyyyy                     | 03082006                   | list index out of range       | month must be in 1..12     | month must be in 1..12     | 2006-03-08 00:00:00        |
| ddmmmyy                      | 03Aug06                    | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| ddmmmyyyy                    | 03Aug2006                  | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| dd-mmm-yy                    | 03-Aug-06                  | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| dd-mmm-yyyy                  | 03-Aug-2006                | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| dmmmyy                       | 3Aug06                     | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| dmmmyyyy                     | 3Aug2006                   | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| d-mmm-yy                     | 3-Aug-06                   | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| d-mmm-yyyy                   | 3-Aug-2006                 | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| d-mmmm-yy                    | 3-August-06                | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| d-mmmm-yyyy                  | 3-August-2006              | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| yymmdd                       | 060803                     | list index out of range       | 2003-06-08 00:00:00        | 2003-06-08 00:00:00        | 2003-06-08 00:00:00        |
| yyyymmdd                     | 20060803                   | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| yy/mm/dd                     | 06/08/03                   | 2003-08-06                    | 2003-06-08 00:00:00        | 2003-06-08 00:00:00        | 2003-06-08 00:00:00        |
| yyyy/mm/dd                   | 2006/08/03                 | day is out of range for month | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| mmddyy                       | 080306                     | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| mmddyyyy                     | 08032006                   | list index out of range       | month must be in 1..12     | month must be in 1..12     | 2006-08-03 00:00:00        |
| mm/dd/yy                     | 08/03/06                   | 2006-03-08                    | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| mm/dd/yyyy                   | 08/03/2006                 | 2006-03-08                    | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| mmm-dd-yy                    | Aug-03-06                  | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| mmm-dd-yyyy                  | Aug-03-2006                | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| yyyy-mm-dd                   | 2006-08-03                 | day is out of range for month | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| weekday, dth mmmm yyyy       | Monday, 3rd of August 2006 | list index out of range       | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        | 2006-08-03 00:00:00        |
| weekday                      | Monday                     | list index out of range       | 2018-08-13 00:00:00        | 2018-08-13 00:00:00        | 2018-08-06 00:00:00        |
| mmm-yy                       | Aug-06                     | list index out of range       | 2018-08-06 00:00:00        | 2018-08-06 00:00:00        | 2018-08-06 00:00:00        |
| yy                           | 06                         | list index out of range       | 2018-08-06 00:00:00        | 2018-08-06 00:00:00        | 2018-06-09 00:00:00        |
| yyyy                         | 2006                       | list index out of range       | 2006-08-09 00:00:00        | 2006-08-09 00:00:00        | 2006-08-09 00:00:00        |
| dd-mmm-yyyy time             | 03-Aug-2006 18:55:30.35    | day is out of range for month | 2006-08-03 18:55:30.350000 | 2006-08-03 18:55:30.350000 | 2006-08-03 18:55:30.350000 |
| yyyy-mm-dd time24 (ODBC Std) | 2006-08-03 18:55:30        | day is out of range for month | 2006-08-03 18:55:30        | 2006-08-03 18:55:30        | 2006-08-03 18:55:30        |
| dd-mmm-yyyy time12           | 03-Aug-2006 6:55:30 pm     | day is out of range for month | 2006-08-03 18:55:30        | 2006-08-03 18:55:30        | 2006-08-03 18:55:30        |
| time24                       | 18:55:30                   | month must be in 1..12        | 2018-08-09 18:55:30        | 2018-08-09 18:55:30        | 2018-08-09 18:55:30        |
| time12                       | 6:55:30 pm                 | day is out of range for month | 2018-08-09 18:55:30        | 2018-08-09 18:55:30        | 2018-08-09 18:55:30        |
| hours                        | 48:55:30                   | month must be in 1..12        | hour must be in 0..23      | hour must be in 0..23      | 5530-04-08 00:00:00        |
| seconds                      | 68538.350                  | list index out of range       | year 68538 is out of range | year 68538 is out of range |                            |

**Comment by wagner-certat** [comment]
> and dateutil.parser raises when parsing ddmmyyyy or mmddyyyy.

Probably because it's ambiguous? Look at the result of dateparser in both cases.

> It also does not deal well with times above 48h

?

**Comment by xmo-odoo** [comment]
> Probably because it's ambiguous? Look at the result of dateparser in both cases.

No, there are plenty of ambiguous datestamps which it's completely fine with, ddmmyy for instance.

> It also does not deal well with times above 48h

I meant 24 (though that's a guess): it's fine with `18:55:30` but not `48:55:30`.

**Comment by pganssle** [comment]
I think the only mildly surprising thing in this list is `mmddyyyy`, which I think is a very uncommon way of expressing a datetime. I vaguely recall discussing this on another issue at some point, but I don't recall the resolution. It seems unlikely that fixing it will cause any problems in the common case, since most datetimes will be within +/- 100 years of today, and that will unambiguously select the last 4 digits as `yyyy`, and the first 4 digits can be disambiguated using `yearfirst` vs `dayfirst`.

As for times with > 24 hours, that's to be expected, since `datetime.time` is bounded between `00:00:00.0000` and `23:59:59.999999`, and real times (when considering leap seconds) have an upper bound of `23:59:60.999999`.

**Comment by xmo-odoo** [comment]
> As for times with > 24 hours, that's to be expected, since datetime.time is bounded between 00:00:00.0000 and 23:59:59.999999, and real times (when considering leap seconds) have an upper bound of 23:59:60.999999.

That's not wrong, but on the other hand dateutil.parse seems to always yield datetimes so parsing a time is more of a delta on the current date, and timedelta is fine with deltas of more than 24h.
